File: day1/solution.py
from pathlib import Path

current_dir = Path(__file__).parent
input_file = Path(current_dir, "input.txt")

lines = input_file.read_text().splitlines()
left = [int(l.split()[0]) for l in lines]
right = [int(l.split()[-1]) for l in lines]


def part1():
    l_s, r_s = list(sorted(left)), list(sorted(right))

    return sum([abs(l_s[i] - r_s[i]) for i in range(len(l_s))])

def part2():
    r_c = {}

    for r in right:
        r_c[r] = r_c.get(r, 0) + 1

    return sum(n * r_c.get(n, 0) for n in left)

print(part1())
print(part2())

File: day1/solution.exs
defmodule Solution do
  def part1(lines) do
    {left, right} = Enum.unzip(lines)

    Enum.zip(Enum.sort(left), Enum.sort(right))
    |> Enum.map(fn {l, r} -> abs(l - r) end)
    |> Enum.sum()
  end

  def part2(lines) do
    r_ref =
      lines
      |> Enum.map(fn {_l, r} -> r end)
      |> Enum.frequencies()

    lines
    |> Enum.map(fn {l, _r} -> l * (r_ref[l] || 0) end)
    |> Enum.sum()
  end
end

lines =
  File.stream!("./input.txt")
  |> Stream.map(&String.split(&1, [" ", "\n"], trim: true))
  |> Stream.map(fn [a, b] -> {String.to_integer(a), String.to_integer(b)} end)

IO.puts(Solution.part1(lines))
IO.puts(Solution.part2(lines))
